packages/expo-application/android/src/main/java/expo/modules/application/ApplicationModule.java

```diff
@@ -0,0 +1,159 @@
+    }
+  }
+
+  @ExpoMethod
+  public void getLastUpdateTimeAsync(Promise promise) {
+    PackageManager packageManager = mContext.getPackageManager();
+    String packageName = mContext.getPackageName();
+    try {
+      PackageInfo info = packageManager.getPackageInfo(packageName, 0);
+      promise.resolve((double)info.lastUpdateTime);
+    } catch (Exception e) {
+      e.printStackTrace();
+      promise.reject("ERR_APPLICATION", "Unable to get last update time of this application.", e);
+    }
+  }
+
+
+  @ExpoMethod
+  public void getInstallReferrerAsync(final Promise promise) {
+
+    final StringBuilder installReferrer = new StringBuilder();
+
+    final InstallReferrerClient referrerClient;
+    referrerClient = InstallReferrerClient.newBuilder(mContext).build();
+    referrerClient.startConnection(new InstallReferrerStateListener() {
+      @Override
+      public void onInstallReferrerSetupFinished(int responseCode) {
+        switch (responseCode) {
+          case InstallReferrerClient.InstallReferrerResponse.OK:
+            // Connection established
+            Log.d("INSTALL_REFERRER", "connection established");
+            try {
+              ReferrerDetails response = referrerClient.getInstallReferrer();
+              installReferrer.append(response.getInstallReferrer());
+              response.getReferrerClickTimestampSeconds();
+              response.getInstallBeginTimestampSeconds();
+              promise.resolve(installReferrer.toString());
+              referrerClient.endConnection();
+            } catch (RemoteException e) {
+              e.printStackTrace();
+              promise.reject("ERR_APPLICATION_INSTALL_REFERRER", "Binding to play store remote invocation error.", e);
+              referrerClient.endConnection();
+            }
+            break;
+          case InstallReferrerClient.InstallReferrerResponse.FEATURE_NOT_SUPPORTED:
+            // API not available on the current Play Store app
+            Log.d("INSTALL_REFERRER", "feature not supported");
+            promise.reject("ERR_APPLICATION_INSTALL_REFERRER", "This feature not supported");
+            referrerClient.endConnection();
+            break;
+          case InstallReferrerClient.InstallReferrerResponse.SERVICE_UNAVAILABLE:
+            // Connection could not be established
+            Log.d("INSTALL_REFERRER", "connection could not be established");
+            promise.reject("ERR_APPLICATION_INSTALL_REFERRER", "Connection could not be established");
+            referrerClient.endConnection();
+            break;
+          default:
+            promise.reject("ERR_APPLICATION_INSTALL_REFERRER", "Connection could not be established");
+            referrerClient.endConnection();
+        }
+      }
+
+      @Override
+      public void onInstallReferrerServiceDisconnected() {
+        // Try to restart the connection on the next request to
+        // Google Play by calling the startConnection() method.
```

Comment:
This seems dangerous if it can cause the JS code to hang waiting for a promise

Reply:
`void onInstallReferrerServiceDisconnected();` is declared in the interface for the `InstallReferrerStateListener` class, so what would you suggest we do?

Reply:
might make sense to save the promise as a member of the class and reject it here

Reply:
this still needs to be resolved, i think -- we definitely want to avoid a hanging promise that never resolves. we may want to keep track of whether or not the promise has been resolved/rejected, and reject it here if it hasn't been